File: ROID-Forum-Server/ROID-Forum-Server/Networking/Message/MessageReader.cs
using System;
using Soly.Utilities.ByteArray;
namespace ROIDForumServer
{
	public class MessageReader(byte[] messageData)
	{
		private uint _currentLoc = 0;
		private readonly ByteArray _byteData = new ByteArray(messageData);
		private readonly int _byteLength = messageData.Length;

		public bool IsAtEndOfData()
		{
			return _byteLength == _currentLoc;
		}

		public bool HasUint8()
		{
			return _currentLoc + 1 <= _byteLength;
		}

		public byte GetUint8()
		{
			var data = _byteData.ReadU8((int)_currentLoc);
			_currentLoc += 1;
			return data;
		}

		public byte PeekUint8()
		{
			return _byteData.ReadU8((int)_currentLoc);
		}

		public bool HasInt8()
		{
			return _currentLoc + 1 <= _byteLength;
		}

		public sbyte GetInt8()
		{
			var data = _byteData.ReadI8((int)_currentLoc);

			_currentLoc += 1;
			return data;
		}

		public sbyte PeekInt8()
		{
			return _byteData.ReadI8((int)_currentLoc);
		}

		public bool HasUint16()
		{
			return _currentLoc + 2 <= _byteLength;
		}

		public ushort GetUint16()
		{
			var data = _byteData.ReadU16((int)_currentLoc, Endianness.BigEndian);
			_currentLoc += 2;
			return data;
		}

		public ushort PeekUint16()
		{
			return _byteData.ReadU16((int)_currentLoc, Endianness.BigEndian);
		}

		public bool HasInt16()
		{
			return _currentLoc + 2 <= _byteLength;
		}

		public short GetInt16()
		{
			var data = _byteData.ReadI16((int)_currentLoc, Endianness.BigEndian);

			_currentLoc += 2;
			return data;

		}

		public short PeekInt16()
		{
			return _byteData.ReadI16((int)_currentLoc, Endianness.BigEndian);
		}

		public bool HasUint32()
		{
			return _currentLoc + 4 <= _byteLength;
		}

		public uint GetUint32()
		{
			var data = _byteData.ReadU32((int)_currentLoc, Endianness.BigEndian);
			_currentLoc += 4;
			return data;
		}

		public uint PeekUint32()
		{
			return _byteData.ReadU32((int)_currentLoc, Endianness.BigEndian);
		}

		public bool HasInt32()
		{
			return _currentLoc + 4 <= _byteLength;
		}

		public int GetInt32()
		{
			var data = _byteData.ReadI32((int)_currentLoc, Endianness.BigEndian);
			_currentLoc += 4;
			return data;
		}

		public int PeekInt32()
		{
			return _byteData.ReadI32((int)_currentLoc, Endianness.BigEndian);
		}

		public double GetFloat64()
		{
			var data = _byteData.ReadF64((int)_currentLoc, Endianness.BigEndian);
			_currentLoc += 8;
			return data;
		}

		public double PeekFloat64()
		{
			return _byteData.ReadF64((int)_currentLoc, Endianness.BigEndian);
		}

		public bool HasFloat64()
		{
			return _currentLoc + 8 <= _byteLength;
		}

		public bool HasFloat32()
		{
			return _currentLoc + 4 <= _byteLength;
		}

		public float GetFloat32()
		{
			var data = _byteData.ReadF32((int)_currentLoc, Endianness.BigEndian);
			_currentLoc += 4;
			return data;
		}

		public float PeekFloat32()
		{
			return _byteData.ReadF32((int)_currentLoc, Endianness.BigEndian);
		}

		public bool HasString()
		{
			if (_currentLoc + 4 > _byteLength)
			{
				return false;
			}
			var length = _byteData.ReadU32((int)_currentLoc, Endianness.BigEndian);
			return _currentLoc + 4 + length <= _byteLength;
		}

		public string GetString()
		{
			var length = _byteData.ReadU32((int)_currentLoc, Endianness.BigEndian);
			_currentLoc += 4;

			var byteArray = new byte[length];
			this._byteData.Read(byteArray, 0, (int)length, (int)_currentLoc);         
			var returnString = System.Text.Encoding.Unicode.GetString(byteArray);
			_currentLoc += length;
			return returnString;
		}

		public string PeekString()
		{
			var length = _byteData.ReadU32((int)_currentLoc, Endianness.BigEndian);

			var byteArray = new byte[length];
			this._byteData.Read(byteArray, 0, (int)length, (int)_currentLoc + 4);         
			return System.Text.Encoding.Unicode.GetString(byteArray);
		}

		public bool HasBinary()
		{
			if (_currentLoc + 4 > _byteLength)
			{
				return false;
			}
			var length = _byteData.ReadU32((int)_currentLoc, Endianness.BigEndian);
			return (_currentLoc + 4 + length) <= _byteLength;
		}

		public byte[] GetBinary()
		{
			var length = _byteData.ReadU32((int)_currentLoc, Endianness.BigEndian);
			_currentLoc += 4;
            var byteArray = new byte[length];
            _byteData.Read(byteArray, 0, (int)length, (int)_currentLoc);  
			_currentLoc += length;
			return byteArray;
		}

		public byte[] PeekBinary()
		{
			var length = _byteData.ReadU32((int)_currentLoc, Endianness.BigEndian);
			var byteArray = new byte[length];
			_byteData.Read(byteArray, 0, (int)length, (int)_currentLoc + 4);
			return byteArray;
		}
        
		public int GetLength()
		{
			return _byteLength;
		}
	}
}

File: ROID-Forum-Web-Client/src/Utility/Message/MessageReader.ts
import {Utility} from "../Utility";

export class MessageReader {
    currentLoc : number;
    byteData : DataView;
    byteLength : number;
    constructor(messageData : ArrayBuffer) {
        this.byteData = new DataView(messageData);
        this.currentLoc = 0;
        this.byteLength = this.byteData.byteLength;
    }

    isAtEndOfData() : boolean {
        return this.byteLength === this.currentLoc;
    }

    hasUint8() : boolean {
        return (this.currentLoc + 1) <= this.byteLength;
    }

    getUint8() : number {
        let data = this.byteData.getUint8(this.currentLoc);
        this.currentLoc += 1;
        return data;
    }

    hasInt8() : boolean {
        return (this.currentLoc + 1) <= this.byteLength;
    }

    getInt8() : number {
        let data = this.byteData.getInt8(this.currentLoc);
        this.currentLoc += 1;
        return data;
    }

    hasUint16() : boolean {
        return (this.currentLoc + 2) <= this.byteLength;
    }

    getUint16() : number {
        let data = this.byteData.getUint16(this.currentLoc, false);
        this.currentLoc += 2;
        return data;
    }

    hasInt16() : boolean {
        return (this.currentLoc + 2) <= this.byteLength;
    }

    getInt16() : number {
        let data = this.byteData.getInt16(this.currentLoc, false);
        this.currentLoc += 2;
        return data;
    }

    hasUint32() : boolean {
        return (this.currentLoc + 4) <= this.byteLength;
    }

    getUint32() : number {
        let data = this.byteData.getUint32(this.currentLoc, false);
        this.currentLoc += 4;
        return data;
    }

    hasInt32() : boolean {
        return (this.currentLoc + 4) <= this.byteLength;
    }

    getInt32() : number {
        let data = this.byteData.getInt32(this.currentLoc, false);
        this.currentLoc += 4;
        return data;
    }

    getFloat64() : number {
        let data = this.byteData.getFloat64(this.currentLoc, false);
        this.currentLoc += 8;
        return data;
    }

    hasFloat64() : boolean {
        return (this.currentLoc + 8) <= this.byteLength;
    }

    hasFloat32() : boolean {
        return (this.currentLoc + 4) <= this.byteLength;
    }

    getFloat32() : number {
        let data = this.byteData.getFloat32(this.currentLoc, false);
        this.currentLoc += 4;
        return data;
    }

    hasString() : boolean {
        if (this.currentLoc + 4 > this.byteLength)
        {
            return false;
        }
        let length = this.byteData.getUint32(this.currentLoc, false);
        return (this.currentLoc + 4 + length) <= this.byteLength;
    }

    getString() : string {
        let length = this.byteData.getUint32(this.currentLoc, false);
        this.currentLoc += 4;
        if (length == 0) {
            return "";
        }
        let stringBuffer = this.byteData.buffer.slice(this.currentLoc, this.currentLoc + length);
        let string = Utility.ArrayBufferToString(stringBuffer);
        this.currentLoc += length;
        return string;
    }

    hasBinary() : boolean {
        if (this.currentLoc + 4 > this.byteLength)
        {
            return false;
        }
        let length = this.byteData.getUint32(this.currentLoc, false);
        return (this.currentLoc + 4 + length) <= this.byteLength;
    }

    getBinary() : ArrayBuffer {
        let length = this.byteData.getUint32(this.currentLoc, false);
        this.currentLoc += 4;
        if (length == 0) {
            return new ArrayBuffer(0);
        }
        let buffer = this.byteData.buffer.slice(this.currentLoc, this.currentLoc + length);
        this.currentLoc += length;
        return buffer;
    }

    getLength() : number {
        return this.byteLength;
    }
}